### PR TITLE
Fix & improve Rise and Fall Time measurement

### PR DESCRIPTION
**Pull request by Marcus10110**
1. The located rise and fall times were nowhere near the center. This now finds the closest edges to the center (and supports where the center point is within an edge)
2. Fixes a bug where the noise would prevent the correct start or edge point from being located. now, once the start end and conditions are located, it will back up and check for another start point that is closer to the end point.

### CODE REVIEW

**Review by huffman — APPROVED** [review]
There are opportunities to condense the code down a bit, but I think we should do this as part of a separate pass. I'd like to finish the test suite, ensure that everything is passing, and then have the freedom to refactor confidently.

File: measurements.py [hunk 1]
```diff
@@ -220,9 +338,24 @@ class TimeMeasurer(AnalogMeasurer):
         measures=[positive_width, negative_width, positive_duty, negative_duty])
 
     def measure_range(self, data: AnalogSpan):
+        if data.clipped:
+            self.positive_width.error = CLIPPED_ERROR
+            self.negative_width.error = CLIPPED_ERROR
+            self.positive_duty.error = CLIPPED_ERROR
+            self.negative_duty.error = CLIPPED_ERROR
+            self.cycle_rms.error = CLIPPED_ERROR
```
Comment (Marcus10110) [inline measurements.py:346]: I'm not sure why width and duty cycle would get these errors but not frequency or period.
RMS makes sense for sure, but the error caused by clipping that affects width and duty cycle would affect frequency/period in the same way, notably that the midpoint voltage will be wrong.

What do you think?

File: measurements.py [hunk 2]
```diff
@@ -220,9 +338,24 @@ class TimeMeasurer(AnalogMeasurer):
...
+
         start_index = int(len(data) / 2)
 
         mid = (data.max + data.min) / 2
+        hysteresis = (data.max - data.min) * FREQUENCY_HYSTERESIS_PCT
+
+        start_value = data[start_index]
+        if (abs(start_value - mid) < hysteresis):
+            lt_index = data.find_lt(mid - hysteresis, start=start_index)
```
Comment (Marcus10110) [inline measurements.py:355]: what do you think  about checking right and left, above and below, and taking the minimum of the 4?

**Review by Marcus10110 — COMMENTED** [review]
2 comments, looks good!

File: measurements.py (hunk 1 above)
Comment (Marcus10110) [reply measurements.py:346]: resolved, this is correct

Comment (huffman) [reply measurements.py:346]: Discussed offline.

File: measurements.py (hunk 2 above)
Comment (huffman) [reply measurements.py:355]: TODO, we will revisit this in a follow-up

**Pull request merged**